forc-plugins/forc-publish/src/credentials.rs

```diff
@@ -0,0 +1,132 @@
+use crate::error::Result;
+use forc_util::user_forc_directory;
+use serde::{Deserialize, Serialize};
+use std::fs;
+use std::io::{self};
+use std::path::PathBuf;
+use toml;
+
+const CREDENTIALS_FILE: &str = "credentials.toml";
+
+#[derive(Serialize, Deserialize)]
+struct Registry {
+    token: String,
+}
+
+#[derive(Serialize, Deserialize)]
+struct Credentials {
+    registry: Registry,
+}
+
+/// Gets the user's auth token.
+/// - First checks CLI arguments.
+/// - Then checks `~/.forc/credentials.toml` inside the `[registry]` section.
+/// - If neither are found, prompts the user and saves it to `credentials.toml`.
+pub fn get_auth_token(
+    opt_token: Option<String>,
+    credentials_dir: Option<PathBuf>,
+) -> Result<String> {
+    if let Some(token) = opt_token {
+        return Ok(token);
+    }
+
+    let credentials_path = credentials_dir
+        .unwrap_or(user_forc_directory())
+        .join(CREDENTIALS_FILE);
+    if let Some(token) = get_auth_token_from_file(&credentials_path)? {
+        return Ok(token);
+    }
+
+    let auth_token =
+        get_auth_token_from_user_input(&credentials_path, io::stdin().lock(), io::stdout())?;
+
+    Ok(auth_token)
+}
+
+// Check if credentials file exists and read from it
+fn get_auth_token_from_file(path: &PathBuf) -> Result<Option<String>> {
+    if path.exists() {
+        let content = fs::read_to_string(path)?;
+        if let Ok(credentials) = toml::from_str::<Credentials>(&content) {
+            return Ok(Some(credentials.registry.token));
+        }
+    }
+    Ok(None)
+}
+
+// Prompt user for input and save to credentials file
+fn get_auth_token_from_user_input<R, W>(
+    credentials_path: &PathBuf,
+    mut reader: R,
+    mut writer: W,
+) -> Result<String>
+where
+    R: io::BufRead,
+    W: io::Write,
+{
+    tracing::info!("Paste your auth token found on https://forc.pub/tokens below: ");
+    writer.flush()?;
+    let mut auth_token = String::new();
+    reader.read_line(&mut auth_token)?;
+    let auth_token = auth_token.trim().to_string();
+
+    // Save the token to the credentials file
+    let credentials = Credentials {
+        registry: Registry {
+            token: auth_token.clone(),
+        },
+    };
+    let toml_content = toml::to_string(&credentials)?;
+    if let Some(parent_path) = credentials_path.parent() {
+        fs::create_dir_all(parent_path)?;
+        fs::write(credentials_path, toml_content)?;
+        tracing::info!("Auth token saved to {}", credentials_path.display());
+    }
```

Comment:
nit: only worth declaring vars and allocating memory if credentials path exists:
```
    if let Some(parent_path) = credentials_path.parent() {
        let credentials = Credentials {
            registry: Registry {
                token: auth_token.clone(),
            },
        };
        
        fs::create_dir_all(parent_path)?;
        fs::write(credentials_path, toml::to_string(&credentials)?)?;
        tracing::info!("Auth token saved to {}", credentials_path.display());
    }
```